File: day02.cpp
#include <iostream>
#include <string>
#include <fstream>

/*
std::string board[] = {"     ",
		       " 123 ",
		       " 456 ",
		       " 789 ",
		       "     "};

*/

/* Board for part 2 */
std::string board[] = {"       ",
		       "   1   ",
		       "  234  ",
		       " 56789 ",
		       "  ABC  ",
		       "   D   ",
		       "       "
};


int main ()
{
  std::string line,code="";
  //int r=1, c=1; // start location for part 1
  int r=3, c=1;  // start location for aprt 

  int dr,dc;
  std::ifstream ifile ("day02.dat");
  if (ifile.is_open()){
    while (getline(ifile,line)){
      for (auto ch  : line){
	dr=0;
	dc=0;
	switch(ch) {
	case 'U':
          dr=-1;
	  break;
	case 'D':
	  dr=1;
	  break;
	case 'L':
	  dc=-1;
	  break;
	case 'R':
	  dc=1;
	  break;
	}
	if (board[r+dr][c+dc] != ' '){
	  r=r+dr;
	  c=c+dc;
	}
	  
        

      }
      code = code + board[r][c];
  
    }
    std::cout << code << std::endl;
  }
  
  return 0;
}

File: day02.clj
(use '[clojure.string :only (join split trim-newline)])
(defn parse-int [s] (Integer/parseInt s))


(def board ["     "
            " 123 "
            " 456 "
            " 789 "
            "     "])


(def b2 [[ \  \  \  \  \  ]
  [\  \1 \2 \3 \ ]
  [\ \4 \5 \6 \ ]
  [\ \7 \8 \9 \ ]
  ])

(def dirs [ '(0 1) '(1 0) '(0 -1) '(-1 0) ])





(def deltas {:R '(0 1)
             :L '(0 -1)
             :U '(-1 -0)
             :D '(1 0)})


(defn get-code
  [[x y num code] move]
  (let [move (keyword (str move))
        [dx dy] (deltas move)
        newx (+ x dx)
        newy (+ y dy)
        
        res (if (= (get-in board [newx newy]) \ )
              [x y (get-in board [x y]) code ]
              [newx newy (get-in board [newx newy]) code])

        ]
    res
    
    ))


(def rawdata (map trim-newline (split(slurp "day02.dat") #"\n")))
(def pos '(2 2 \5 ""))


(-> (reduce (fn [[x y num code] line]
          (let [ [nx ny nnum ncode] (reduce get-code [x y num code] line)
                ]
            [nx ny num (str nnum code)]
            )
              )pos rawdata)
    (nth 3)
    clojure.string/reverse
    )

